rsciio/zspy/_api.py

```diff
@@ -109,6 +105,12 @@ def _get_object_dset(group, data, key, chunks, **kwds):
             dtype = data[test_ind].compute().dtype
         else:
             dtype = data[test_ind].dtype
+        if np.issubdtype(dtype, str):
+            size_of_char = np.dtype("U1").itemsize
+            # Go through the whole array to find the largest string type
+            for index in np.ndindex(data.shape):
+                if data[index].dtype.itemsize // size_of_char > dtype.itemsize:
+                    dtype = data[index].dtype
```

Comment:
How does this work with lazy arrays of strings?  I don't think that this should work? If it does it won't be very efficient as you have to look at every single index. 

We don't __have__ to support saving lazy arrays of strings but we should at least document it...  Part of me is just tempted to pre allocate something like 128 characters but as soon as I say that then something is going to come up where I want more than 128.

Reply:
Yes, I have a look at the feasibility of the alternative approach mentioned in https://github.com/hyperspy/rosettasciio/pull/213#issuecomment-1902630331, because if this is possible, this would be done in the `flatten`/`unflatten` functions and therefore be lazy friendly.

Reply:
I think to do that you would basically have to store everything as 1 long string:

For example 

```["a", "b", "cd", "e"] --> "abcde"``` and then also save an int array `[1,1,2,1]`

You could also use a deliminator:
```["a", "b", "cd", "e"] --> "a-b-cd-e"```


The flatten function:

```
def flatten_strings(arr):
    string_lengths = np.empty(arr.shape, dtype=object)
    flat_strings= np.empty(arr.shape, dtype=object)
    for i in np.ndindex(arr):
        string_lengths[I] = np.char.str_len(arr)
        flat_strings[i] =  "".join( arr[I])
    return string_lengths, flat_strings
```

Reply:
Note that 2-d ragged arrays of strings might be more of a headache so I'm okay with sticking to the 1d case.